modules/delegates.js

```diff
@@ -353,112 +353,146 @@ __private.decryptPassphrase = function(encryptedPassphrase, password) {
  * @returns {setImmediateCallback} cb, err
  * @todo Add description for the params
  */
-__private.checkDelegates = function(publicKey, votes, state, cb, tx) {
+__private.checkDelegates = function(senderPublicKey, votes, state, cb, tx) {
 	if (!Array.isArray(votes)) {
 		return setImmediate(cb, 'Votes must be an array');
 	}
 
-	modules.accounts.getAccount(
-		{ publicKey },
-		(err, account) => {
-			if (err) {
-				return setImmediate(cb, err);
-			}
+	// Converting publicKeys to addresses because module.accounts do not support fetching multiple accounts by publicKeys
+	const actionWithAddresses = votes.map(vote => {
+		const action = vote[0];
+		let voteAddress;
 
-			if (!account) {
-				return setImmediate(cb, 'Account not found');
+		try {
+			if (action !== '+' && action !== '-') {
+				throw 'Invalid math operator';
 			}
+			const votePublicKey = vote.slice(1);
 
-			const delegates =
-				state === 'confirmed' ? account.delegates : account.u_delegates;
-			const existingVotes = Array.isArray(delegates) ? delegates.length : 0;
-			let additions = 0;
-			let removals = 0;
-
-			async.eachSeries(
-				votes,
-				(action, cb) => {
-					const math = action[0];
-
-					if (math === '+') {
-						additions += 1;
-					} else if (math === '-') {
-						removals += 1;
-					} else {
-						return setImmediate(cb, 'Invalid math operator');
-					}
-
-					const publicKey = action.slice(1);
-
-					try {
-						Buffer.from(publicKey, 'hex');
-					} catch (e) {
-						library.logger.error(e.stack);
-						return setImmediate(cb, 'Invalid public key');
-					}
-
-					modules.accounts.getAccount(
-						{ publicKey, isDelegate: 1 },
-						(err, account) => {
-							if (err) {
-								return setImmediate(cb, err);
-							}
-
-							if (!account) {
-								return setImmediate(cb, 'Delegate not found');
-							}
-
-							if (
-								math === '+' &&
-								(delegates != null && delegates.indexOf(publicKey) !== -1)
-							) {
-								return setImmediate(
-									cb,
-									`Failed to add vote, delegate "${
-										account.username
-									}" already voted for`
-								);
-							}
-
-							if (
-								math === '-' &&
-								(delegates === null || delegates.indexOf(publicKey) === -1)
-							) {
-								return setImmediate(
-									cb,
-									`Failed to remove vote, delegate "${
-										account.username
-									}" was not voted for`
-								);
-							}
-
-							return setImmediate(cb);
-						},
-						tx
-					);
-				},
-				err => {
-					if (err) {
-						return setImmediate(cb, err);
-					}
-
-					const totalVotes = existingVotes + additions - removals;
+			voteAddress = modules.accounts.generateAddressByPublicKey(votePublicKey);
+		} catch (e) {
+			library.logger.error(e.stack);
```

Comment:
also some additional message about source

Reply:
e.stack is very descriptive. For example, 
> x = new Error('database error')
> x.stack
'Error: database error\n    at repl:1:5\n    at sigintHandlersWrap (vm.js:22:35)\n    at sigintHandlersWrap (vm.js:73:12)\n    at ContextifyScript.Script.runInThisContext (vm.js:21:12)\n    at REPLServer.defaultEval (repl.js:340:29)\n    at bound (domain.js:280:14)\n    at REPLServer.runBound [as eval] (domain.js:293:12)\n    at REPLServer.<anonymous> (repl.js:539:10)\n    at emitOne (events.js:101:20)\n    at REPLServer.emit (events.js:188:7)'
It gives all the details about the error.

Reply:
Yes I agree with @4miners we need to have a descriptive error message here, as we are doing everywhere else in application. May something like: 

> Invalid vote format found in transaction

Reply:
If there is an exception thrown from https://github.com/LiskHQ/lisk/pull/2189/files#diff-9d5cd48f1ba4ec6b2770fd44edc90323R369, we will have a message `Invalid math operator`. And also, `modules.accounts.generateAddressByPublicKey` throws `Invalid public key: ${publicKey}` in case of invalid publicKey. This is as descriptive as it gets.